relm4-examples/libadwaita/examples/sudoku_solver.rs

```diff
@@ -0,0 +1,197 @@
+use gtk::glib::Sender;
+use gtk::prelude::{BoxExt, ButtonExt, GtkWindowExt, OrientableExt};
+use relm4::adw::prelude::WidgetExt;
+use relm4::adw::traits::AdwApplicationWindowExt;
+use relm4::factory::positions::GridPosition;
+use relm4::factory::{FactoryPrototype, FactoryVec};
+use relm4::{adw, gtk, send, AppUpdate, Model, RelmApp, WidgetPlus, Widgets};
+use sudoku::Sudoku;
+
+#[derive(Debug)]
+enum AppMsg {
+    Solve,
+    Clear,
+    Update(usize, u8),
+}
+
+#[derive(Clone, Debug)]
+struct Number {
+    value: u8,
+}
+
+struct AppModel {
+    values: FactoryVec<Number>,
+}
+
+impl AppModel {
+    fn empty() -> Self {
+        Self {
+            values: FactoryVec::from_vec(vec![Number { value: 0 }; 81]),
+        }
+    }
+
+    fn to_line_string(&self) -> String {
+        let mut res = String::new();
+        for v in self.values.iter() {
+            let string = match v.value {
+                0 => String::from("."),
+                v @ 1..=9 => format!("{v}"),
+                v => panic!("{v} is not defined"),
+            };
+            res.push_str(string.as_str())
+        }
+        res
+    }
+
+    fn from_string(&mut self, input: String) {
+        for (i, v) in input.chars().enumerate() {
+            if v != '.' {
+                let value = v.to_string().parse().unwrap();
+                if let Some(v) = self.values.get_mut(i) {
+                    v.value = value;
+                }
+            }
+        }
+    }
+}
+
+impl Model for AppModel {
+    type Msg = AppMsg;
+    type Widgets = AppWidgets;
+    type Components = ();
+}
+
+impl AppUpdate for AppModel {
+    fn update(&mut self, msg: AppMsg, _components: &(), _sender: Sender<AppMsg>) -> bool {
+        match msg {
+            AppMsg::Solve => {
+                let content = self.to_line_string();
+                println!("Start solving with {content}");
+                let sudoku = Sudoku::from_str_line(content.as_str()).unwrap();
+
+                if let Some(solution) = sudoku.solve_one() {
+                    self.from_string(solution.to_string());
+                }
+            }
+            AppMsg::Clear => {
+                // Why does this not work
+                // self.values.clear();
```

Comment:
Do you know, why this is not working? What would be the best way to set the initial values?

Reply:
If `self.values.clear()` doesn't clear the list and also destroys all widgets that'd be a bug. Can you open an issue if that's the case? Nothing complicated, just as reminder.

Reply:
But after your suggestions, I think, the workaround is maybe not too bad, What do you think.

Reply:
Actually, if I see this currently, the workaround isn't a workaround after all. But there's a kinda big rework of Relm4 on the way and I might change factories to make them work with such code.